contrib/ruby/ext/trilogy-ruby/cext.c

```diff
@@ -224,6 +270,66 @@ struct rb_trilogy_wait_args {
     int rc;
 };
 
+#ifdef TRILOGY_RB_IO_WAIT
+static int _cb_ruby_wait(trilogy_sock_t *sock, trilogy_wait_t wait)
+{
+    struct trilogy_ctx *ctx = sock->user_data;
+    struct timeval *timeout = NULL;
+    int wait_flag = 0;
+
+    switch (wait) {
+    case TRILOGY_WAIT_READ:
+        timeout = &sock->opts.read_timeout;
+        wait_flag = RUBY_IO_READABLE;
+        break;
+
+    case TRILOGY_WAIT_WRITE:
+        timeout = &sock->opts.write_timeout;
+        wait_flag = RUBY_IO_WRITABLE;
+        break;
+
+    case TRILOGY_WAIT_CONNECT:
+        // wait for connection to be writable
+        timeout = &sock->opts.connect_timeout;
+        if (timeout->tv_sec == 0 && timeout->tv_usec == 0) {
+            // We used to use the write timeout for this, so if a connect timeout isn't configured, default to that.
+            timeout = &sock->opts.write_timeout;
+        }
+        wait_flag = RUBY_IO_WRITABLE;
+        break;
+
+    case TRILOGY_WAIT_HANDSHAKE:
+        // wait for handshake packet on initial connection
+        timeout = &sock->opts.connect_timeout;
+        wait_flag = RUBY_IO_READABLE;
+        break;
+
+    default:
+        return TRILOGY_ERR;
+    }
+
+    if (ctx->io == Qnil) {
+        VALUE io = rb_io_open_descriptor(rb_cIO, trilogy_sock_fd(sock), FMODE_EXTERNAL, RUBY_Qnil, RUBY_Qnil, NULL);
+        RB_OBJ_WRITE(ctx->self, &ctx->io, io);
+    }
+
+    if (timeout->tv_sec == 0 && timeout->tv_usec == 0) {
+        timeout = NULL;
+    }
+
+    VALUE result = rb_io_wait(ctx->io, RB_INT2NUM(wait_flag), rb_fiber_scheduler_make_timeout(timeout));
```

Comment:
rb_fiber_scheduler_make_timeout() is not implemented yet on TruffleRuby: https://github.com/search?q=repo%3Aoracle%2Ftruffleruby%20rb_fiber_scheduler_make_timeout&type=code
Yet `#if defined(HAVE_RB_IO_WAIT) && defined(RB_IO_OPEN_DESCRIPTOR) && defined(HAVE_RUBY_FIBER_SCHEDULER_H)` should be true on TruffleRuby.

So could you check for `have_func("rb_fiber_scheduler_make_timeout")` in extconf.rb?

(BTW [that function](https://github.com/ruby/ruby/blob/97e6ad49a4604e7e4ca04da2aaafc63cbd5d29d8/scheduler.c#L270-L278) uses `0.000001f` (32-bit float) shouldn't it be `0.000001` (64-bit double) instead?)

Reply:
Yes we can fix those issues, thanks for pointing them out.

Reply:
Here is the fix to prefer double math: https://github.com/ruby/ruby/pull/12803

Reply:
Okay, fixed.